### ISSUE TITLE
LE - Still need SAN even with redirect

### ISSUE THREAD

**Issue by kazdegroot** [omega8cc/boa#964]
Currently the LE functionality _explicitly_ doesn't add SAN records for aliases when redirection is enabled. This creates two problems:

1. Aliases that redirect to the 'canonical' domain now still serve errors, cause their DNS names aren't incorporated into the certificate
2. (At least for me) the CN chosen for the certificate was the main name of the site instead of the redirect target; Resulting in my site redirecting from a working page to one with a nice certificate error.

Nr. 1 seems like a deliberate choice, that I really don't understand, as it breaks the redirect functionality while 2 seems to be a bug.

**Comment by omega8cc** [comment]
ad 2
This has been fixed in the last 24 hours, I believe -- you can either upgrade octopus again or run:

`rm -f /var/xdrago/log/hosting_le.fixed*`
`rm -f /var/xdrago/conf/provision_hosting_le.drush.inc`
`bash /var/xdrago/clear.sh`

Then disable SSL on the affected site, wait until Verify is run, and then enable it again.

ad 1
This is expected, because with redirect enabled, real SSL cert will list only the main site name or redirection target name, if used, hence the warning.

There is no workaround for this, unless you will disable redirection, because letsencrypt will try to verify every alias we are trying to add to the SAN names, and when it hits redirect on any alias, it will never accept this alias name and entire challenge will fail (no cert generated nor updated), so we have to exclude them automatically, if redirection is enabled.

Theoretically we could trick ACME by temporarily disabling redirection, running ACME challenge, and then enabling redirection again. But we would have to do it daily/weekly -- and basically every time the internal Verify tasks are run (and they can be run even three times in a row during Migrate/Rename tasks).

This makes all attempt to avoid ACME requirement rather hopeless, I think.

**Comment by kazdegroot** [comment]
Wrt. 1; I have some other (non BOA) servers where I got around that issue by doing the following:
```
server {
        listen 80;
        server_name sub1.example.com
          sub2.example.com
          sub2.example.com;

        location /.well-known/acme-challenge/ {
        	default_type "text/plain";
        	alias /etc/nginx/le-challenges/;
        	try_files $uri =404;
        }

        location / {
                return 302 https://$host$request_uri;
        }
}
```
Is there some reason such an approach wouldn't work here?

**Comment by omega8cc** [comment]
All vhosts are always automatically re-generated when you run Verify task, and you can't create conflicting server {} instances, or nginx will use only the first loaded and ignore the other (reporting in the error log that it noticed duplicate server_name's.)

Once Aegir created redirect server {} entries, you can't do anything about it, nor automate overriding this -- either you want redirections enabled, or not, right?

**Comment by omega8cc** [comment]
It is a classic chicken/egg syndrome situation.

**Comment by kazdegroot** [comment]
I'll go and complain in the provision issue queue to support modifying the redirect server {} blocks.
I hope you agree this severely limits the usability of the redirect functionality of the redirects in combination with LE, as the alternative domains are still without an valid certificate.

**Comment by omega8cc** [comment]
There is this old feature request to separate aliases and redirects: https://www.drupal.org/node/815190

**Comment by omega8cc** [comment]
I think that we could do this pretty easily in our Provision fork. We would need to just add the ACME path location within every redirection server {}, just above the `return 301` line.

But currently there is no way to extend these server {} parts via hooks.

**Comment by omega8cc** [comment]
Fixed after a number of commits in: https://github.com/omega8cc/hosting_le plus: https://github.com/omega8cc/provision/commit/920a8738a47ae00fb5d2038d3856c7689f646b25

**Comment by omega8cc** [comment]
Fix added as hotfix to `BOA-3.1.0` -- just run the octopus forced upgrade to get it applied.

**Comment by omega8cc** [comment]
Now let's update the docs.
